test/higher_order_ops/test_with_effects.py

```diff
@@ -0,0 +1,191 @@
+                return (res,)
+
+        inputs = (torch.randn(3),)
+
+        # Without functionalization, print should just appear in the graph directly
+        gm = make_fx(M())(*inputs)
+        FileCheck().check_count("torch.ops.aten._print.default", 2, exactly=True).run(
+            gm.code
+        )
+
+        # With functionalization, it should appear wrapped with tokenize()
+        gm, gs = aot_export_module(M(), inputs, trace_joint=False)
+        self.assertExpectedInline(
+            str(gm.code).strip(),
+            """\
+def forward(self, arg0_1, arg1_1):
+    with_effects = torch._higher_order_ops.effects.with_effects(arg0_1, torch.ops.aten._print.default, 'moo');  arg0_1 = None
+    getitem = with_effects[0];  with_effects = None
+    add = torch.ops.aten.add.Tensor(arg1_1, arg1_1);  arg1_1 = None
+    with_effects_1 = torch._higher_order_ops.effects.with_effects(getitem, torch.ops.aten._print.default, 'moo');  getitem = None
+    getitem_2 = with_effects_1[0];  with_effects_1 = None
+    return [getitem_2, add]""",
+        )
+        self.assertEqual(len(gs.input_tokens), 1)
+        self.assertEqual(len(gs.output_tokens), 1)
+
+    def test_torchbind_custom_op(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.attr = torch.classes._TorchScriptTesting._Foo(10, 20)
+
+            def forward(self, x):
+                return (x + torch.ops._TorchScriptTesting.takes_foo(self.attr, x),)
+
+        with enable_torchbind_tracing():
+            gm, gs = aot_export_module(M(), (torch.ones(2, 3),), trace_joint=False)
+
+        self.assertExpectedInline(
+            str(gm.code).strip(),
+            """\
+def forward(self, arg0_1, arg1_1):
+    _tensor_constant0 = self._tensor_constant0
+    with_effects = torch._higher_order_ops.effects.with_effects(arg0_1, torch.ops._TorchScriptTesting.takes_foo.default, _tensor_constant0, arg1_1);  arg0_1 = _tensor_constant0 = None
+    getitem = with_effects[0]
+    getitem_1 = with_effects[1];  with_effects = None
+    add = torch.ops.aten.add.Tensor(arg1_1, getitem_1);  arg1_1 = getitem_1 = None
+    return [getitem, add]""",  # noqa: B950
+        )
+        self.assertEqual(len(gs.input_tokens), 1)
+        self.assertEqual(len(gs.output_tokens), 1)
+
+    def test_print_with_buffer_mutations(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer("buf", torch.ones(3))
+
+            def forward(self, x):
+                torch.ops.aten._print("moo")
+                res = x + x
+                self.buf.add_(res)
+                res = self.buf + x
+                torch.ops.aten._print("moo")
+                return (res,)
+
+        inputs = (torch.randn(3),)
+
+        # With functionalization, it should appear wrapped with tokenize()
+        gm, gs = aot_export_module(M(), inputs, trace_joint=False)
```

Comment:
I'm not familiar with this API -- what is gs?

Reply:
graph signature -- it's this type: https://fburl.com/code/32p134uh